### PR TITLE
Generalize `window` to support open/closed ranges.

### PR DESCRIPTION
**Pull request by ryzhyk**
This cannot be merged until the compiler has been adjusted to use the new API.

The `window` operator used to treat the key range as closed on the left,
open on the right: [start..end).

This commit generalizes the operator to support any combinations of
open/closed intervals.

Is this a user-visible change (yes/no): no

### CODE REVIEW

File: crates/dbsp/src/operator/dynamic/time_series/window.rs [hunk 1]
```diff
@@ -219,8 +291,12 @@ where
             // If the window shrunk, retract values that dropped off the right end of the
             // window.
             if &end1 < end0 {
-                trace_cursor.seek_key(&end1);
-                while trace_cursor.key_valid() && trace_cursor.key() < end0 {
+                seek_after_end(&mut trace_cursor, &end1, self.right_inclusive);
+
+                //trace_cursor.seek_key(&end1);
```
Comment (mihaibudiu) [inline crates/dbsp/src/operator/dynamic/time_series/window.rs:296]: delete?

File: crates/dbsp/src/trace/mod.rs [hunk 2]
```diff
@@ -489,6 +489,30 @@ where
     /// [`Batch::recede_to`].
     fn recede_to(&mut self, frontier: &Self::Time);
 
+    /// Returns elements from `self` that satisfy a predicate.
+    fn filter(&self, predicate: &dyn Fn(&Self::Key, &Self::Val) -> bool) -> Self
```
Comment (mihaibudiu) [inline crates/dbsp/src/trace/mod.rs:493]: I think this should be pub
We need more functions like this

**Review by mihaibudiu — APPROVED** [review]
I will push a commit that fixes the compiler

File: crates/dbsp/src/trace/mod.rs (hunk 2 above)
Comment (ryzhyk) [reply crates/dbsp/src/trace/mod.rs:493]: It's a trait method, so it's automatically `pub`.

**Comment by mihaibudiu** [comment]
Fixes #2303

**Pull request merged**